### ISSUE TITLE
2.3.0 dramatically increases time to process training data files

### ISSUE THREAD

**Issue by erohmensing** [RasaHQ/rasa#8012]
**Rasa version**:
2.3.0

**Python version**:
from `rasa/rasa` docker images

**Operating system** (windows, osx, ...):
from `rasa/rasa` docker images

**Issue**:
When training a relatively large model in docker using the same machine/docker specs, the process of preparing the training data
(`2021-02-19 18:06:14 DEBUG    rasa.shared.nlu.training_data.loading  - Training data format of '<filename>' is 'rasa_yml'.`) went from taking, from the first of these logs to the last: 

*2.2.10*: 10:45:34 - 10:57:07: about 13 minutes 
*2.3.0*: 12:22:55 - 18:41:05: about 4 hours, 20 minutes

**Command or request that led to error**:
```
docker run -d -v $(pwd):/app rasa/rasa:2.2.10-full train --debug
docker run -d -v $(pwd):/app rasa/rasa:2.3.0-full train --debug
```

It should be noted that each of the NLU files are processed at least 3 times (this is how many times guess_format prints). Core files are only processed once. 

Looking at the 3 different times this loop of guessing format is called, here is the change in processing time: 

Loop 1: 
*2.2.10*: 10:45:34 - 10:45:34: one second
*2.3.0*: 12:22:55 - 12:22:55: one second

Loop 2: 
*2.2.10*: 10:56:04 - 10:56:30: 26 seconds 
*2.3.0*: 14:03:32 - 15:59:37: 1 hour 56 mins

Loop 3 (after the first `Added 827 training data examples from the story training data` log):
*2.2.10*: 10:56:32 - 10:57:07:  35 seconds 
*2.3.0*: 16:01:04 - 18:41:05: 2 hours 40 minutes 

So it seems like maybe some very slow processing code that is only called the 2nd and 3rd time this data is processed was added to the code. Either that, or maybe there is some cleanup issue that doesn't affect the first processing step but sticks around to affect the latter.

**Comment by erohmensing** [comment]
Suspected it might be due to the update to the `pykwalify` library. After making a simple docker container based off of 2.3.0 where I explicitly installed `pykwalify==1.7.0`, these were the results: 

Loop 1:
11:14:52 - 11:14:52 - one second

Loop 2:
11:25:04 - 11:25:30 - 26 seconds 

Loop 3 (after the first Added 827 training data examples from the story training data log):
11:25:32 - 11:26:04 - 32 seconds

I'd suspect that in the latter 2 loops, we do some YAML validation that is not part of the `guess_format` method that doesn't exist in the first loop. For now, I'd recommend pinning `pykwalify` until we can look further into how to handle this update smoothly (or open an issue in their repo depending on what we find)

**Comment by erohmensing** [comment]
Results on 2.3.1: 

Loop 1:
13:02:24 - 13:02:24 - one second

Loop 2:
13:14:09 - 13:14:39 - 30 seconds

Loop 3 (after the first Added 827 training data examples from the story training data log):
13:14:41 - 13:15:24 - 43 seconds

Anyway, looks like this was already fixed in 2.3.1 (https://github.com/RasaHQ/rasa/pull/7970), thanks for going on this ride with me 🤦🏻
